lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```
@@ -0,0 +1,105 @@
+   * @param timeoutUnits The unit of time for the timeoutPeriod parameter.
+   */
+  public MultiAsyncToSync(final long timeoutPeriod, final TimeUnit timeoutUnits) {
+    this.timeoutPeriod = timeoutPeriod;
+    this.timeoutUnits = timeoutUnits;
+  }
+
+  /**
+   * Put the caller to sleep on a specific release identifier.
+   * @param identifier The identifier required to awake the caller via the release() method.
+   * @return
+   */
+  public boolean block(final long identifier) {
+    ConditionVariable call;
+    synchronized (this) {
+      // Get an condition variable to block the calling thread.
+      if (sleeperMap.containsKey(identifier)) {
+        // This should never happen as message identifiers are unique.
+        LOG.log(Level.SEVERE, "Duplicate identifier in RPC map");
+      }
+      if (freeQueue.isEmpty()) {
+        freeQueue.addLast(new ConditionVariable(timeoutPeriod, timeoutUnits));
+      }
+      call = freeQueue.getFirst();
+      sleeperMap.put(identifier, call);
+    }
+
+    LOG.log(Level.INFO, "Putting caller to sleep on identifier [{0}]", identifier);
+    // Put the call to sleep until the ack comes back.
+    boolean timeoutOccurred = call.waitForSignal();
+    if (timeoutOccurred) {
+      synchronized (this) {
+        call = sleeperMap.remove(identifier);
+        freeQueue.addLast(call);
+      }
+      LOG.log(Level.INFO, "Caller sleeping on identifier [{0}] timed out", identifier);
+    }
+    return timeoutOccurred;
+  }
+
+  /**
+   * Wake the caller sleeping on the specific identifier.
+   * @param identifier The message identifier of the caller who should be released.
+   */
+  public void release(final long identifier) {
+    ConditionVariable call;
+    synchronized (this) {
+      // Get the associated call object.
+      call = sleeperMap.remove(identifier);
+      if (call == null) {
+        LOG.log(Level.SEVERE, "Unknown message identifier [{0}]", identifier);
```

When a timeout occurs, we remove the ConditionVariable from the map because we do not know if the asynchronous processing operation will ever complete. However, if the asynchronous processing does complete after the time out, it will call release() and get this error. The timeout test case actually exhibits this behavior. I guess we can throw the exception and if the application considers some number of lost messages is acceptable they can catch it and keep going. 

So I have created an exception class for this case.